### ISSUE TITLE
isWrapped Issue - Never True On Last Line

### ISSUE THREAD

**Issue by thagh05t** [xtermjs/xterm.js#2472]
Hey,

First off, I love this project, it opens up a lot of doors for development for me.

but.. (not directly related to my current hurdle, but i'd like to know if there's an easier way)

I am trying to implement xterm.js in an angular app as a terminal emulator. I have been writing methods to handle different key events such as:

- Backspace
- Enter
- Delete
- Home
- End

It appears that I have to handle the cursor position and data buffers for all of these, is there an easier way? For example.

When someone hits Backspace, the cursor should move backwards and the previous character from starting cursor position should be removed from the line. That's the easy part, but any subsequent lines that are wrapped then have to be adjusted accordingly. I am currently doing this by reading each line into a temporary buffer, clearing that line, and then setting the cursor position at the beginning of the first line and reprinting the modified temp buffer which includes all of those lines.

I really do not mind this, but I do see a large memory consumption sometimes when dealing with a few wrapped lines. If there is an easier or more efficient way, could you please point me in the right direction??? Similar concepts were used to achieve the other inputs.

My Current Hurdle...

I am using the fit add-on for obvious reasons. Entering characters, backspace, enter, etc. all work up until I get to the last line (this.cols -1). When I get to the last line, if the text wraps the isWrapped property (this.buffer.getLine(this.buffer.cursorY).isWrapped)) doesn't reflect its true state. My guess is because the current line technically cannot be wrapped as it is a new line in the buffer? js/ts is not my native language so please be gentle....

Any help or example will be greatly appreciated, I am sure this is just my lack of understanding of the component, but on the off chance that it is a bug, I am speaking up...

**Comment by Tyriar** [comment]
So for your first query, I think you're doing mostly the right thing. For a readline-like system you would need to redraw everything that's changed, I believe you should not be relying on the terminal's line wrapping through and instead using `\r\n` or positioning the cursor. When you use just `\r` to wrap a line then the terminal's "reflow" feature will attempt to wrap lines when things are resized (note also that the cursor line is never reflowed).

`isWrapped` indicates whether the line wraps onto the next line, because of this is will never be true for the last line in the buffer.

**Comment by thagh05t** [comment]
Thanks for the quick reply!

I do not quite understand what you mean by:

> I believe you should not be relying on the terminal's line wrapping through and instead using \r\n...

I have been using the built in buffer (albeit it says its experimental?) to get what i need done so far. Basically, if the character wrapping happens without intervention and I just use the isWrapped attribute to let me know where I should stop reading if I am inserting or deleting a character from within a line. I go to the first line that !isWrapped and iterate through all rows -1. I take the result and reprint it. I am sure that's what your understanding was anyways, but wanted to reiterate. 

Are you suggesting that I do not use the isWrapped attribute and base my temp buffer on something else? I started off creating and using my own linebuffer, but ran into some other issues. I suppose I could abandon the isWrapped attribute and use something else... How would I determine what row is the actual start of the command if the command has wrapped n rows already?

**Comment by Tyriar** [comment]
@thagh05t it sounds like you're trying to write an app to run in the terminal by looking at properties on xterm.js, this isn't typically how it's done and would also mean that it only works in xterm.js. The buffer API itself is meant to enable terminals to support cool features, a example of this could be scanning the buffer and extracting any links that were printed onto it. Another example is being able to save the buffer state and restore it.

Normally your app would run via a pty where a process is launched on the other side, this separation is important to allow an application to run on many different terminals.

You can check out this example of node-pty/xterm.js running under electron as an example: https://github.com/microsoft/node-pty/tree/master/examples/electron, when launched it will open your terminal shell, from there you can run terminal applications. I suggest you target running a terminal application and look into things like ncurses or readline implementations for the language you're targeting.

> How would I determine what row is the actual start of the command if the command has wrapped n rows already?

I think shells do this by remembering which lines they have written to and rewrite them when changes are made.

**Comment by thagh05t** [comment]
Thanks for the input.

I suppose I may be misusing this a bit. At the moment I have no intentions of hooking it up to a backend, only readline-like operations and capturing the input command and having angular handle it. the application is only meant to emulate at this point.

I have found other jquery implementation of terminals, would this be a better option for my use case considering I am not looking to use xterm.js to its real potential?

**Comment by Tyriar** [comment]
@thagh05t this might be what you're after:

https://github.com/wavesoft/local-echo
https://medium.com/swlh/local-echo-xterm-js-5210f062377e

**Comment by thagh05t** [comment]
Thanks for all your help, I appreciate you taking the time to answer a non-bug question.

I did find local-echo, but couldn't it to work after the first line return... likely just me not understanding how it works. I decided to just try to do it myself since I need the practice anyways. 

I think I will run with a jquery terminal for now as I got it up and running within 5 minutes, but I have plans to revisit xterm.js for something more advanced that actually has a backend. Keep up the good work guys!
